File: src/main/java/main/utils/apis/robertify/imagebuilders/AbstractImageBuilder.java
package main.utils.apis.robertify.imagebuilders;

import lombok.SneakyThrows;
import main.constants.ENV;
import main.main.Config;
import me.duncte123.botcommons.web.WebUtils;
import okhttp3.OkHttpClient;
import org.apache.http.NameValuePair;
import org.apache.http.client.utils.URIBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;
import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

public abstract class AbstractImageBuilder {
    final Logger logger = LoggerFactory.getLogger(AbstractImageBuilder.class);

    private final URIBuilder uri;
    private final OkHttpClient httpClient;
    private final WebUtils webUtils;

    @SneakyThrows
    AbstractImageBuilder(ImageType imageType) {
        final long DEFAULT_TIMEOUT = 2;

        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .readTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .writeTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .build();
        this.webUtils = WebUtils.ins;

        final var segments = new ArrayList<>(List.of("api", "images"));
        segments.addAll(imageType.getSegments());
        this.uri = new URIBuilder(Config.get(ENV.ROBERTIFY_WEB_HOSTNAME))
                .setPathSegments(segments);
    }

    protected void addQuery(ImageQueryField key, String value) {
        uri.addParameter(key.toString(), value);
    }

    protected String findQuery(ImageQueryField key) {
        return uri.getQueryParams()
                .stream()
                .filter(param -> param.getName().equalsIgnoreCase(key.toString()))
                .findFirst()
                .map(NameValuePair::getValue)
                .orElse(null);
    }

    @SneakyThrows
    public InputStream build() throws ImageBuilderException {
        final var url = new URL(uri.build().toString());
        logger.debug("Built image URL:\n{}", url);

        try {
            return httpClient.newCall(webUtils.prepareGet(url.toString()).build())
                    .execute()
                    .body()
                    .byteStream();
        } catch (SocketTimeoutException | ConnectException e) {
            throw new ImageBuilderException(e);
        } catch (IOException e) {
            logger.error("Unexpected error", e);
        }
        return null;
    }

    public static String getRandomFileName() {
        return UUID.randomUUID() + ".png";
    }

    public interface ImageQueryField {}
}

File: src/main/kotlin/main/utils/api/robertify/imagebuilders/AbstractImageBuilder.kt
package main.utils.api.robertify.imagebuilders

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import main.main.Config
import main.utils.api.robertify.imagebuilders.models.ImageQueryField
import okhttp3.OkHttpClient
import org.apache.http.client.utils.URIBuilder
import org.slf4j.LoggerFactory
import java.io.IOException
import java.io.InputStream
import java.net.ConnectException
import java.net.SocketTimeoutException
import java.util.UUID
import java.util.concurrent.CompletionException
import java.util.concurrent.TimeUnit

abstract class AbstractImageBuilder protected constructor(imageType: ImageType) {
    companion object {
        private val logger = LoggerFactory.getLogger(Companion::class.java)
        private const val DEFAULT_TIMEOUT = 2L
        val RANDOM_FILE_NAME: String
            get() = "${UUID.randomUUID()}.png"
    }

    private val httpClient = HttpClient(CIO) {
        engine {
            requestTimeout = DEFAULT_TIMEOUT * 1000
        }
    }
    private var uri: URIBuilder

    init {
        val segments = listOf("api", "images").toMutableList()
        segments.addAll(imageType.getSegments())
        uri = URIBuilder(Config.ROBERTIFY_WEB_HOSTNAME)
            .setPathSegments(segments)
    }

    protected fun addQuery(key: ImageQueryField, value: String) =
        uri.addParameter(key.toString(), value)

    protected fun findQuery(key: ImageQueryField) =
        uri.queryParams
            .find { it.name.equals(key.toString(), true) }
            ?.value

    open suspend fun build(): InputStream? {
        val url = uri.build().toURL().toString()

        return try {
            httpClient.get(url).body<InputStream>()
        } catch (e: SocketTimeoutException) {
            throw ImageBuilderException(cause = e)
        } catch (e: ConnectException) {
            throw ImageBuilderException(cause = e)
        } catch (e: CompletionException) {
            throw ImageBuilderException(cause = e)
        } catch (e: IOException) {
            logger.error("Unexpected error", e)
            null
        }
    }
}